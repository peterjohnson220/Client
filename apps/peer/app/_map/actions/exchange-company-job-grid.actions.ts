import { Action } from '@ngrx/store';
import { GridDataResult } from '@progress/kendo-angular-grid';

export const LOAD_EXCHANGE_COMPANY_JOBS  = '[Peer Main/Map/Exchange Company Jobs] Load Exchange Company Jobs';
export const LOAD_EXCHANGE_COMPANY_JOBS_SUCCESS  = '[Peer Main/Map/Exchange Company Jobs] Load Exchange Company Jobs Success';
export const LOAD_EXCHANGE_COMPANY_JOBS_ERROR = '[Peer Main/Map/Exchange Company Jobs] Load Exchange Company Jobs Error';
export const LOAD_EXCHANGE_COMPANY_JOBS_IDS  = '[Peer Main/Map/Exchange Company Jobs] Load Exchange Company Jobs Ids';
export const LOAD_EXCHANGE_COMPANY_JOBS_IDS_SUCCESS  = '[Peer Main/Map/Exchange Company Jobs] Load Exchange Company Jobs Ids Success';
export const LOAD_EXCHANGE_COMPANY_JOBS_IDS_ERROR = '[Peer Main/Map/Exchange Company Jobs] Load Exchange Company Jobs Ids Error';

export class LoadExchangeCompanyJobs implements Action {
  readonly type = LOAD_EXCHANGE_COMPANY_JOBS;
}

export class LoadExchangeCompanyJobsSuccess implements Action {
  readonly type = LOAD_EXCHANGE_COMPANY_JOBS_SUCCESS;

  constructor(public payload: GridDataResult, public exchangeJobIds: number[]) {}
}

export class LoadExchangeCompanyJobsError implements Action {
  readonly type = LOAD_EXCHANGE_COMPANY_JOBS_ERROR;
}

export class LoadExchangeCompanyJobsIds implements Action {
  readonly type = LOAD_EXCHANGE_COMPANY_JOBS_IDS;
}

export class LoadExchangeCompanyJobsIdsSuccess implements Action {
  readonly type = LOAD_EXCHANGE_COMPANY_JOBS_IDS_SUCCESS;

  constructor(public exchangeJobToCompanyJobIds: number[]) {}
}

export class LoadExchangeCompanyJobsIdsError implements Action {
  readonly type = LOAD_EXCHANGE_COMPANY_JOBS_IDS_ERROR;
}

export type Actions
  = LoadExchangeCompanyJobs
  | LoadExchangeCompanyJobsSuccess
  | LoadExchangeCompanyJobsError
  | LoadExchangeCompanyJobsIds
  | LoadExchangeCompanyJobsIdsSuccess
  | LoadExchangeCompanyJobsIdsError;
