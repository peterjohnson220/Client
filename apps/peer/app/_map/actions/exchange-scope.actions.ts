import { Action } from '@ngrx/store';

export const OPEN_SAVE_EXCHANGE_SCOPE_MODAL = '[Peer Main/Map] Open Save Exchange Scope Modal';
export const CLOSE_SAVE_EXCHANGE_SCOPE_MODAL = '[Peer Main/Map] Close Save Exchange Scope Modal';

export class OpenSaveExchangeScopeModal implements Action {
  readonly type = OPEN_SAVE_EXCHANGE_SCOPE_MODAL;
}

export class CloseSaveExchangeScopeModal implements Action {
  readonly type = CLOSE_SAVE_EXCHANGE_SCOPE_MODAL;
}

export type Actions
  = OpenSaveExchangeScopeModal
  | CloseSaveExchangeScopeModal;
