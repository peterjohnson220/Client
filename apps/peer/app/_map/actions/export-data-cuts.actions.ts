import { Action } from '@ngrx/store';

export const EXPORT_DATA_CUTS = '[Peer Main/Map/Export Data Cuts] Export Data Cuts';
export const EXPORT_DATA_CUTS_NEW = '[Peer Main/Map/Export Data Cuts New] Export Data Cuts New';
export const EXPORT_DATA_CUTS_SUCCESS = '[Peer Main/Map/Export Data Cuts] Export Data Cuts Success';
export const EXPORT_DATA_CUTS_ERROR = '[Peer Main/Map/Export Data Cuts] Export Data Cuts Error';
export const OPEN_EXPORT_DATA_CUTS_MODAL = '[Peer Main/Map/Export Data Cuts] Open Export Data Cuts Modal';
export const CLOSE_EXPORT_DATA_CUTS_MODAL = '[Peer Main/Map/Export Data Cuts] Close Export Data Cuts Modal';
export const SELECT_RATE = '[Peer Main/Map/Export Data Cuts] Select Rate';
export const SELECTED_RATE_PERSISTED = '[Peer Main/Map/Export Data Cuts] Selected Rate Persisted';
export const SELECT_WEIGHTING_TYPE = '[Peer Main/Map/Export Data Cuts] Select Weighting Type';
export const SELECTED_WEIGHTING_TYPE_PERSISTED = '[Peer Main/Map/Export Data Cuts] Selected Weighting Type Persisted';

export class ExportDataCuts implements Action {
  readonly type = EXPORT_DATA_CUTS;

  constructor(public payload: {selectedRate: string, scopes: string[], exportCurrentMap: boolean, selectedWeightingType: string}) {}
}

export class ExportDataCutsNew implements Action {
  readonly type = EXPORT_DATA_CUTS_NEW;

  constructor(public payload: {selectedRate: string, scopes: string[], exportCurrentMap: boolean, selectedWeightingType: string}) {}
}

export class ExportDataCutsSuccess implements Action {
  readonly type = EXPORT_DATA_CUTS_SUCCESS;
}

export class ExportDataCutsError implements Action {
  readonly type = EXPORT_DATA_CUTS_ERROR;
}

export class OpenExportDataCutsModal implements Action {
  readonly type = OPEN_EXPORT_DATA_CUTS_MODAL;
}

export class CloseExportDataCutsModal implements Action {
  readonly type = CLOSE_EXPORT_DATA_CUTS_MODAL;
}

export class SelectRate implements Action {
  readonly type = SELECT_RATE;

  constructor(public payload: {newRate: string}) {}
}

export class SelectedRatePersisted implements Action {
  readonly type = SELECTED_RATE_PERSISTED;
}

export class SelectWeightingType implements Action {
  readonly type = SELECT_WEIGHTING_TYPE;

  constructor(public payload: {newWeightingType: string}) {}
}

export class SelectedWeightingTypePersisted implements Action {
  readonly type = SELECTED_WEIGHTING_TYPE_PERSISTED;
}

export type Actions
  = ExportDataCuts
  | ExportDataCutsNew
  | ExportDataCutsSuccess
  | ExportDataCutsError
  | OpenExportDataCutsModal
  | CloseExportDataCutsModal
  | SelectRate
  | SelectedRatePersisted
  | SelectWeightingType
  | SelectedWeightingTypePersisted;
