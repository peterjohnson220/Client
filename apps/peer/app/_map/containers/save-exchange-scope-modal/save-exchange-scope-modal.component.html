<pf-modal-form
  [size]="'md'"
  [title]="'Save Exchange Scope'"
  [primaryButtonText]="'Save'"
  [submitting]="upsertingExchangeScope$ | async"
  [isOpen$]="saveExchangeScopeModalOpen$"
  [formGroup]="saveExchangeScopeForm"
  (onSubmit)="handleFormSubmit()"
  (onDismiss)="handleModalDismissed()">
  <ng-container form-content>
    <input type="text" class="form-control" placeholder="Name" formControlName="exchangeScopeName" id="exchange-scope-name"
           [readonly]="upsertingExchangeScope$ | async"
           [pfValidatable]="true"
           pfFocus>
    <ng-container *ngIf="exchangeScopeNameControl.errors">
      <div *ngIf="exchangeScopeNameControl.errors.exchangeScopeNameExists" class="invalid-feedback">
        {{ exchangeScopeNameControl.errors.exchangeScopeNameExists }}
      </div>
    </ng-container>
    <span *ngIf="!exchangeScopeNameControl?.errors?.exchangeScopeNameExists" class="pf-placeholder"></span>
  </ng-container>
</pf-modal-form>
