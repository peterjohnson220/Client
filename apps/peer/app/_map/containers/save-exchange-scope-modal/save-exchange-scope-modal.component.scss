pf-multi-select {
  margin-top: 1rem;
  display: block;
}
