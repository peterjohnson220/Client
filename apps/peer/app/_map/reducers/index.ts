import { createSelector, createFeatureSelector } from '@ngrx/store';

// Import root app reducer
import { IFeatureGridState } from 'libs/core/reducers/grid.reducer';
import * as fromRoot from 'libs/state/state';
import * as fromGridReducer from 'libs/core/reducers/grid.reducer';

import * as fromExchangeCompanyJobsReducer from '../reducers/exchange-company-job-grid.reducer';
import * as fromExportDataCutsReducer from '../reducers/export-data-cuts.reducer';

// Feature area state
export interface PeerMapState {
  exportDataCuts: fromExportDataCutsReducer.State;
  exchangeCompanyJobsGrid: IFeatureGridState<fromExchangeCompanyJobsReducer.State>;
}

// Extend root state with feature area state
export interface State extends fromRoot.State {
  peer_map: PeerMapState;
}

// Feature area reducers
export const reducers = {
  exportDataCuts: fromExportDataCutsReducer.reducer,
  exchangeCompanyJobsGrid: fromExchangeCompanyJobsReducer.reducer
};

// Select Feature Area
export const selectFeatureAreaState = createFeatureSelector<PeerMapState>('peer_map');

// Feature Selectors
export const selectExchangeCompanyJobsState = createSelector(
  selectFeatureAreaState,
  (state: PeerMapState) => state.exchangeCompanyJobsGrid
);

export const selectExportDataCutsState = createSelector(
  selectFeatureAreaState,
  (state: PeerMapState) => state.exportDataCuts
);

// Export Data Cuts
export const getDataCutsExporting = createSelector(
  selectExportDataCutsState,
  fromExportDataCutsReducer.getExporting
);

export const getDataCutsExportingError = createSelector(
  selectExportDataCutsState,
  fromExportDataCutsReducer.getExportingError
);

export const getExportDataCutsModalOpen = createSelector(
  selectExportDataCutsState,
  fromExportDataCutsReducer.getExportDataCutsModalOpen
);

export const getExportDataCutsModalCurrencies = createSelector(
  selectExportDataCutsState,
  fromExportDataCutsReducer.getCurrencies
);

// Exchange Company Jobs Grid
export const selectExchangeCompanyJobsFeatureState = createSelector(
  selectExchangeCompanyJobsState,
  (state: IFeatureGridState<fromExchangeCompanyJobsReducer.State>) => state.feature
);

export const selectExchangeCompanyJobsGridState = createSelector(
  selectExchangeCompanyJobsState,
  (state: IFeatureGridState<fromExchangeCompanyJobsReducer.State>) => state.grid
);

export const {
  selectAll: getExchangeCompanyJobs
} = fromExchangeCompanyJobsReducer.adapter.getSelectors(selectExchangeCompanyJobsFeatureState);

export const getExchangeCompanyJobsLoading = createSelector(
  selectExchangeCompanyJobsFeatureState,
  fromExchangeCompanyJobsReducer.getLoading
);

export const getExchangeCompanyJobsLoadingError = createSelector(
  selectExchangeCompanyJobsFeatureState,
  fromExchangeCompanyJobsReducer.getLoadingError
);

export const getExchangeCompanyJobsTotal = createSelector(
  selectExchangeCompanyJobsFeatureState,
  fromExchangeCompanyJobsReducer.getTotal
);

export const getExchangeCompanyJobsGridSelections = createSelector(
  selectExchangeCompanyJobsGridState,
  fromGridReducer.getGridSelections
);

export const getExchangeCompanyJobsGridSelectAllState = createSelector(
  selectExchangeCompanyJobsGridState,
  fromGridReducer.getGridSelectAllState
);

export const getExchangeCompanyJobsGridState = createSelector(
  selectExchangeCompanyJobsGridState,
  fromGridReducer.getGridState
);

export const getExchangeCompanyJobsGridData = createSelector(
  getExchangeCompanyJobs,
  getExchangeCompanyJobsTotal,
  (data, total) => {
    return { data: data, total: total };
  }
);

export const getExchangeCompanyJobsAllIds = createSelector(
  selectExchangeCompanyJobsFeatureState,
  fromExchangeCompanyJobsReducer.getAllIds
);
