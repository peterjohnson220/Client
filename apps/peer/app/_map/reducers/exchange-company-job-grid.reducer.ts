import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';

import { createGridReducer } from 'libs/core/reducers/grid.reducer';
import { GridTypeEnum } from 'libs/models/common';

import * as fromExchangeCompanyJobActions from '../actions/exchange-company-job-grid.actions';
import { ExchangeCompanyJob } from '../models';

export interface State extends EntityState<ExchangeCompanyJob> {
  loading: boolean;
  loadingError: boolean;
  loadingAllIds: boolean;
  loadingAllIdsError: boolean;
  total: number;
  allIds: number[];
}

export const adapter: EntityAdapter<ExchangeCompanyJob> = createEntityAdapter<ExchangeCompanyJob>({
  selectId: (exchangeCompanyJob: ExchangeCompanyJob) => exchangeCompanyJob.ExchangeJobToCompanyJobId
});

const initialState: State = adapter.getInitialState({
  loading: false,
  loadingError: false,
  loadingAllIds: false,
  loadingAllIdsError: false,
  allIds: [],
  total: 0
});


// Reducer function
export function reducer(state, action) {
  return createGridReducer(
    GridTypeEnum.ExchangeCompanyJob,
    (featureState = initialState,  featureAction: fromExchangeCompanyJobActions.Actions): State => {
      switch (featureAction.type) {
        case fromExchangeCompanyJobActions.LOAD_EXCHANGE_COMPANY_JOBS: {
          return {
            ...featureState,
            loading: true,
            loadingError: false
          };
        }
        case fromExchangeCompanyJobActions.LOAD_EXCHANGE_COMPANY_JOBS_SUCCESS: {
          const exchangeJobMappings: ExchangeCompanyJob[] = featureAction.payload.data;
          const exchangeJobIdsInScope = featureAction.exchangeJobIds;
          const exchangeJobMappingsFinal = exchangeJobMappings.map(ejm => {
            return {
              ...ejm,
              IsInMapScope: exchangeJobIdsInScope.includes(ejm.ExchangeJobId)
            };
          });
          return {
            ...adapter.removeAll(featureState),
            ...adapter.addAll(exchangeJobMappingsFinal, featureState),
            total: featureAction.payload.total,
            loading: false
          };
        }
        case fromExchangeCompanyJobActions.LOAD_EXCHANGE_COMPANY_JOBS_ERROR: {
          return {
            ...featureState,
            loading: false,
            loadingError: true
          };
        }
        case fromExchangeCompanyJobActions.LOAD_EXCHANGE_COMPANY_JOBS_IDS: {
          return {
            ...featureState,
            loadingAllIds: true,
            loadingAllIdsError: false
          };
        }
        case fromExchangeCompanyJobActions.LOAD_EXCHANGE_COMPANY_JOBS_IDS_SUCCESS: {
          return {
            ...featureState,
            allIds: action.exchangeJobToCompanyJobIds,
            loadingAllIds: false,
            loadingAllIdsError: false
          };
        }
        case fromExchangeCompanyJobActions.LOAD_EXCHANGE_COMPANY_JOBS_IDS_ERROR: {
          return {
            ...featureState,
            loadingAllIdsError: true
          };
        }
        default: {
          return featureState;
        }
      }
    }, {
      take: 10
    })(state, action);
}

// Selector functions
export const getLoading = (state: State) => state.loading;
export const getLoadingError = (state: State) => state.loadingError;
export const getTotal = (state: State) => state.total;
export const getAllIds = (state: State) => state.allIds;
