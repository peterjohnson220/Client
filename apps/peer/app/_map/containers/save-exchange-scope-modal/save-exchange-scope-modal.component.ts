import { Component, EventEmitter, Input, OnDestroy, OnInit, Output } from '@angular/core';
import { AbstractControl, AsyncValidatorFn, FormBuilder, FormGroup, ValidationErrors, Validators } from '@angular/forms';

import { Store, select } from '@ngrx/store';
import { Observable, Subscription, timer } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';

import { PfValidators } from 'libs/forms/validators/pf-validators';
import { ExchangeScopeApiService } from 'libs/data/payfactors-api';
import { Exchange } from 'libs/models/peer';
import { GenericKeyValue, GenericMenuItem } from 'libs/models/common';
import * as fromExchangeExplorerReducer from 'libs/features/peer/exchange-explorer/reducers';

import * as fromSaveExchangeScopeActions from '../../actions/save-exchange-scope.actions';
import * as fromPeerMapReducer from '../../reducers';

@Component({
  selector: 'pf-save-exchange-scope-modal',
  templateUrl: './save-exchange-scope-modal.component.html',
  styleUrls: ['./save-exchange-scope-modal.component.scss']
})
export class SaveExchangeScopeModalComponent implements OnInit, OnDestroy {
  @Input() exchange: Exchange;
  @Output() upsertExchangeScopeEvent = new EventEmitter();

  upsertingExchangeScope$: Observable<boolean>;
  upsertingExchangeScopeError$: Observable<boolean>;
  saveExchangeScopeModalOpen$: Observable<boolean>;
  parentPayMarketOptionsLoading$: Observable<boolean>;
  parentPayMarketOptions$: Observable<GenericKeyValue<number, string>[]>;

  saveExchangeScopeModalOpenSub: Subscription;
  parentPayMarketOptionsSub: Subscription;

  saveExchangeScopeForm: FormGroup;
  selectedParentPayMarketOptions: GenericMenuItem[];
  parentPayMarketOptions: GenericMenuItem[];
  validatingScopeName = false;

  constructor(
    private store: Store<fromPeerMapReducer.State>,
    private fb: FormBuilder,
    private exchangeScopeApiService: ExchangeScopeApiService) {
    this.upsertingExchangeScope$ = this.store.pipe(select(fromExchangeExplorerReducer.getExchangeScopeUpserting));
    this.upsertingExchangeScopeError$ = this.store.pipe(select(fromExchangeExplorerReducer.getExchangeScopeUpsertingError));
    this.saveExchangeScopeModalOpen$ = this.store.pipe(select(fromPeerMapReducer.getSaveExchangeScopeModalOpen));
    this.parentPayMarketOptions$ = this.store.pipe(select(fromPeerMapReducer.getSaveExchangeScopeParentPayMarketOptions));
    this.parentPayMarketOptionsLoading$ = this.store.pipe(select(fromPeerMapReducer.getSaveExchangeScopeParentPayMarketOptionsLoading));
    this.createForm();
  }

  get exchangeScopeNameControl() { return this.saveExchangeScopeForm.get('exchangeScopeName'); }
  get exchangeScopeDescriptionControl() { return this.saveExchangeScopeForm.get('exchangeScopeDescription'); }
  get isDefaultScopeControl() { return this.saveExchangeScopeForm.get('isDefaultScope'); }

  get descriptionPlaceholder(): string {
    return `Add a brief description about the Exchange Scope you are creating...`;
  }

  createForm(): void {
    this.saveExchangeScopeForm = this.fb.group({
      'exchangeScopeName': ['', [PfValidators.required, Validators.minLength(3)], [this.exchangeScopeNameValidator()]],
      'exchangeScopeDescription': [''],
      'isDefaultScope': [false]
    });
  }

  handleFormSubmit(): void {
    const selectedPayMarketIds = !!this.selectedParentPayMarketOptions?.length ? this.selectedParentPayMarketOptions.map(mi => +mi.Value) : [];
    this.upsertExchangeScopeEvent.emit({
      Name: this.exchangeScopeNameControl.value,
      Description: this.exchangeScopeDescriptionControl.value,
      IsDefault: this.isDefaultScopeControl.value,
      CompanyPayMarketIdsToDefaultFor: selectedPayMarketIds
    });
  }

  handleModalDismissed(): void {
    this.store.dispatch(new fromSaveExchangeScopeActions.CloseSaveExchangeScopeModal);
    this.selectedParentPayMarketOptions = [];
  }

  exchangeScopeNameValidator(): AsyncValidatorFn {
    return (control: AbstractControl): Observable<ValidationErrors> => {
      const exchangeScopeName = control.value;
      const exchangeExists = !!this.exchange;
      const exchangeId = exchangeExists ? this.exchange.ExchangeId : 0;
      const exchangeName = exchangeExists ? this.exchange.ExchangeName : '';
      this.validatingScopeName = true;
      return timer(500).pipe(
        switchMap(() => {
        return this.exchangeScopeApiService
          .validateExchangeScopeName(exchangeId, exchangeScopeName).pipe(
            map(isValid => {
              this.validatingScopeName = false;
              return isValid !== true ? {
                exchangeScopeNameExists: 'The ' + exchangeName + ' exchange already has a scope by this name'
              } : null;
            }));
      }));
    };
  }

  // Life-cycle Events
  ngOnInit(): void {
    this.saveExchangeScopeModalOpenSub = this.saveExchangeScopeModalOpen$.subscribe(open => {
      if (!!open) {
        this.store.dispatch(new fromSaveExchangeScopeActions.LoadParentPayMarkets());
      }
    });

    this.parentPayMarketOptionsSub = this.parentPayMarketOptions$.subscribe(ppmo => {
      this.parentPayMarketOptions = ppmo?.map(pm => ({DisplayName: pm.Value, Value: pm.Key.toString(), IsSelected: false}));
    });
  }

  ngOnDestroy(): void {
    this.saveExchangeScopeModalOpenSub.unsubscribe();
    this.parentPayMarketOptionsSub.unsubscribe();
  }
}
