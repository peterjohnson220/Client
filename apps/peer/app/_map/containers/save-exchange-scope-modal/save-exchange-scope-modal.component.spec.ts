import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { TestBed, ComponentFixture } from '@angular/core/testing';

import { combineReducers, Store, StoreModule } from '@ngrx/store';

import * as fromRootState from 'libs/state/state';
import * as fromLibsPeerExchangeExplorerReducers from 'libs/features/peer/exchange-explorer/reducers';
import {
  generateMockUpsertExchangeExplorerScopeRequest, UpsertExchangeExplorerScopeRequest} from 'libs/models/peer/requests/upsert-exchange-scope-request.model';
import { ExchangeScopeApiService } from 'libs/data/payfactors-api/peer';

import * as fromPeerMapReducer from '../../reducers';
import * as fromExchangeScopeActions from '../../actions/save-exchange-scope.actions';

import { SaveExchangeScopeModalComponent } from './save-exchange-scope-modal.component';

describe('Peer - Map - Save Exchange Scope Modal', () => {
  const mockUpsertExchangeScopeRequest: UpsertExchangeExplorerScopeRequest = generateMockUpsertExchangeExplorerScopeRequest();
  let fixture: ComponentFixture<SaveExchangeScopeModalComponent>;
  let instance: SaveExchangeScopeModalComponent;
  let store: Store<fromRootState.State>;

  // Configure Testing Module for before each test
  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          ...fromRootState.reducers,
          peer_map: combineReducers(fromPeerMapReducer.reducers),
          feature_peer_exchangeExplorer: combineReducers(fromLibsPeerExchangeExplorerReducers.reducers)
        }),
        ReactiveFormsModule
      ],
      declarations: [
        SaveExchangeScopeModalComponent
      ],
      providers: [
        {
          provide: ExchangeScopeApiService,
          useValue: { validateExchangeScopeName: jest.fn() }
        }
      ],
      // Shallow Testing
      schemas: [ NO_ERRORS_SCHEMA ]
    });

    store = TestBed.inject(Store);

    jest.spyOn(store, 'dispatch');

    fixture = TestBed.createComponent(SaveExchangeScopeModalComponent);
    instance = fixture.componentInstance;
  });

  it('should show the form text and no error messages on init', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch an CloseSaveExchangeScopeModal action when handleModalDismissed is called', () => {
    const action = new fromExchangeScopeActions.CloseSaveExchangeScopeModal();

    instance.handleModalDismissed();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should clear selectedParentPayMarketOptions when handleModalDismissed is called', () => {
    const action = new fromExchangeScopeActions.CloseSaveExchangeScopeModal();

    instance.handleModalDismissed();

    expect(instance.selectedParentPayMarketOptions).toStrictEqual([]);
  });

  it('should emit an upsertExchangeScopeEvent event with exchangeScopeName when handleFormSubmit is called', () => {
    instance.exchangeScopeNameControl.setValue(mockUpsertExchangeScopeRequest.ExchangeScopeDetails.ExchangeScopeName);
    instance.exchangeScopeDescriptionControl.setValue(mockUpsertExchangeScopeRequest.ExchangeScopeDetails.ExchangeScopeDescription);

    jest.spyOn(instance.upsertExchangeScopeEvent, 'emit');

    instance.handleFormSubmit();

    expect(instance.upsertExchangeScopeEvent.emit).toHaveBeenCalledWith({
      Name: mockUpsertExchangeScopeRequest.ExchangeScopeDetails.ExchangeScopeName,
      Description: mockUpsertExchangeScopeRequest.ExchangeScopeDetails.ExchangeScopeDescription,
      IsDefault: !!mockUpsertExchangeScopeRequest.ExchangeScopeDetails.IsDefault,
      CompanyPayMarketIdsToDefaultFor: mockUpsertExchangeScopeRequest.ExchangeScopeDetails.CompanyPayMarketIdsToDefaultFor
    });
  });
});
