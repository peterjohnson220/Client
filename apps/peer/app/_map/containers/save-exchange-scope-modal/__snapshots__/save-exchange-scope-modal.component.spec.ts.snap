// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Peer - Map - Save Exchange Scope Modal should show the form text and no error messages on init 1`] = `
<pf-save-exchange-scope-modal
  exchangeScopeApiService={[Function Object]}
  fb={[Function FormBuilder]}
  saveExchangeScopeForm={[Function FormGroup]}
  saveExchangeScopeModalOpen$={[Function Store]}
  store={[Function Store]}
  upsertExchangeScopeEvent={[Function EventEmitter]}
  upsertingExchangeScope$={[Function Store]}
  upsertingExchangeScopeError$={[Function Store]}
  validatingScopeName="false"
>
  <pf-modal-form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    title="Create Exchange Scope"
  >
    
    <input
      class="form-control ng-untouched ng-pristine ng-invalid"
      formcontrolname="exchangeScopeName"
      id="exchange-scope-name"
      ng-reflect-name="exchangeScopeName"
      pffocus=""
      placeholder="Name"
      type="text"
    />
    
    
    
    
    
    <span
      class="pf-placeholder"
    />
  </pf-modal-form>
</pf-save-exchange-scope-modal>
`;
